### ISSUE TITLE
Discussion: When to stop supporting Python versions

### ISSUE THREAD

**Issue by duckinator** [duckinator/emanate#63]
This issue is to document what Python versions are shipped by various operating systems, to get a feel for what we need to support.

Python branch EOL dates are explained at https://devguide.python.org/#status-of-python-branches

## Notes

1. Python 2 is explicitly unsupported
2. In general, I plan to support a version until it reaches EOL.
3. Python 3.5 EOL is just over a year away, but some distros don't ship anything newer.

## OSes/distros and Python versions

OS/distro                       | OS/distro EOL  | Python version
------------------------------|---------------------|--------------------
ArchLinux                     | N/A                   | 3.7
Debian 9 (Stretch)        | 2020                 | 3.5
Debian 10 (Buster)       | 2022                 | 3.7
Fedora 29                    | F31 release + 1 month (estimated: Nov 2019) | 3.7 
Fedora 30                    | F32 release + 1 month (estimated: Jun 2020)  | 3.7
OpenSUSE Leap 15.0 | Nov 2019           | 3.6
OpenSUSE Leap 15.1 | Nov 2020           | 3.6
Ubuntu 18.04               | ???                    | 3.6.5
Ubuntu 19.04               | ???                   | 3.7.3
DragonFly BSD 5         | ???                   | 3.6.8
FreeBSD 11                 | 2021                  | 3.6.8
FreeBSD 12                 | 2024                  | 3.6.8
OpenBSD 6                  | ???                   | 3.6
NetBSD 8                     | ???                    | 3.4, 3.5, 3.6, 3.7 as separate packages

Notes about this table:

1. Fedora policy is "two releases in the future, plus a month", which is on average 13 months. https://fedoraproject.org/wiki/Fedora_Release_Life_Cycle#Maintenance_Schedule

**Comment by duckinator** [comment]
So, the table is now complete, and the only still-supported semi-mainstream Linux or BSD variant I could find that doesn't provide Python 3.6 or newer is Debian 9 (Stretch).

However, Buster is only going to become Debian stable _this month_ (July 2019).

I'm leaning towards either:

1. revisiting this when Buster is stable + has been out for a while. (maybe 3-5 months?)
2. just saying fuck it and waiting until Python 3.5 hits EOL.

@nbraud what're your thoughts?

**Comment by nbraud** [comment]
@duckinator Personally, I would currently be inconvenienced if we dropped Py3.5 support: some of the machines I `emanate` my dotfiles on, but don't administer, aren't going to be updated to Buster soon (or any other Py3.6+ platform).

From a maintenance perspective, it doesn't seem to be a big burden: the only commits I found related to supporting older versions are 9081e382242a9f7d27d8ee9027b4c891af1b3f1a, and 24f26880fd139a0ff0c41127a7b943f051bc18e5.

So I would lean towards (2), though nothing prevents us from rethinking that decision later this year.
